Common/Securities/SecurityCacheProvider.cs

```diff
@@ -63,26 +62,22 @@ public SecurityCache GetSecurityCache(Symbol symbol)
                 {
                     // we didn't find the underlying, lets keep track of the underlying symbol which might get added in the future.
                     // else when it is added, we would have to go through existing Securities and find any which use it as underlying
-                    if (_relatedSymbols.ContainsKey(symbol.Underlying.ID))
-                    {
-                        _relatedSymbols[symbol.Underlying.ID].Add(symbol);
-                    }
-                    else
-                    {
-                        _relatedSymbols[symbol.Underlying.ID] = new ConcurrentBag<Symbol> { symbol };
-                    }
+                    _relatedSymbols.AddOrUpdate(symbol.Underlying,
+                        id => new ConcurrentBag<Symbol> { symbol },
+                        (id, bag) => { bag.Add(symbol); return bag; }
+                    );
                 }
             }
-            else if(_relatedSymbols.ContainsKey(symbol.ID))
+            else
             {
-                // if we are here it means we added a symbol which is an underlying of some existing custom symbols
                 ConcurrentBag<Symbol> customSymbols;
-                if (_relatedSymbols.TryRemove(symbol.ID, out customSymbols))
+                if (_relatedSymbols.TryRemove(symbol, out customSymbols))
                 {
+                    // if we are here it means we added a symbol which is an underlying of some existing custom symbols
                     foreach (var customSymbol in customSymbols)
                     {
-                        Security customSecurity;
-                        if (_securityProvider.TryGetValue(customSymbol, out customSecurity))
+                        var customSecurity = _securityProvider.GetSecurity(customSymbol);
+                        if (customSecurity != null)
```

Comment:
no need to change -- I think at some point I added a `TryGetValue` extension method to maintain the beauty of the try pattern.